### PR TITLE
[dune] Simple rule to generate Stdlib's documentation.

### PR DESCRIPTION
**Pull request by ejgallego**
Ideally this will be handled by Dune's native library support, but
this could be useful for the likes of #9648.

I am not sure what should be done w.r.t. style files.

### CODE REVIEW

File: doc/stdlib/dune [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+(rule
+  (targets html)
+  (deps
+   (source_tree %{project_root}/theories)
+   (source_tree %{project_root}/plugins)
+   (:header %{project_root}/doc/common/styles/html/coqremote/header.html)
+   (:footer %{project_root}/doc/common/styles/html/coqremote/footer.html))
+ (action
+  (progn
+   (bash "mkdir -p html")
```
Comment (rgrinberg) [inline doc/stdlib/dune]: A bit curious why you need bash here. `(run mkdir -p html)` doesn't work?

File: doc/stdlib/dune [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+   (bash "%{bin:coqdoc} -q -d html --with-header %{header} --with-footer %{footer} --multi-index --html -g -R %{project_root}/theories Coq -R %{project_root}/plugins Coq $(find %{project_root}/theories %{project_root}/plugins -name *.v)"))))
```
Comment (rgrinberg) [inline doc/stdlib/dune]: Same here. It should be possible to do this without bash. Just bind the dependencies in theories and plugins to a name. If that doesn't work, I think we should have a bug report for that.

Comment (ejgallego) [reply doc/stdlib/dune]: If I bind the dependencies I'll get all the files in the tree, how can I filter just the `.v` ones?

I had to use bash here due to quoting issues when inlining the `find` call.

File: doc/stdlib/dune (hunk 1 above)
Comment (ejgallego) [reply doc/stdlib/dune]: Good point, thanks I never know if this kind of commands are shell builtins or not.

File: doc/stdlib/dune (hunk 2 above)
Comment (rgrinberg) [reply doc/stdlib/dune]: I see, so actually, the feature that you need here is a recursive glob. Basically, you want to write:

```
(deps (coq_files (glob_files theories/**/*.v)))
```

I'm not entirely sure if we have a ticket for this, but the feature was mentioned a few times before. I'll look it up.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -340,7 +340,7 @@ doc:refman:dune:
   <<: *dune-ci-template
   variables:
     <<: *dune-ci-template-vars
-    DUNE_TARGET: refman-html
+    DUNE_TARGET: refman-html stdlib-html
```
Comment (SkySkimmer) [inline .gitlab-ci.yml]: Any reason to do it in the same job vs adding another job?

File: doc/stdlib/dune [hunk 4]
```diff
@@ -0,0 +1,17 @@
+; This is an ad-hoc rule to easy the migration, it should be handled
+; natively by Dune in the future.
+(rule
+  (targets html)
+  (deps
+   (source_tree %{project_root}/theories)
+   (source_tree %{project_root}/plugins)
+   (:header %{project_root}/doc/common/styles/html/coqremote/header.html)
+   (:footer %{project_root}/doc/common/styles/html/coqremote/footer.html))
```
Comment (SkySkimmer) [inline doc/stdlib/dune]: Doesn't coqdoc read glob files or some such thing?

File: .gitlab-ci.yml (hunk 3 above)
Comment (ejgallego) [reply .gitlab-ci.yml]: Efficiency. One documentation job seems enough IMO. In particular the deploy job needs to download this. That should change if we split the stdlib build off tho.

File: doc/stdlib/dune (hunk 4 above)
Comment (ejgallego) [reply doc/stdlib/dune]: Good point, on the current setup they will be there as part of the artifacts but indeed we should add the package dependency on.

When native support comes the right rule will be generated tho.

File: doc/stdlib/dune (hunk 2 above)
Comment (rgrinberg) [reply doc/stdlib/dune]: Here's the ticket if anyone is curious: https://github.com/ocaml/dune/issues/1866

**Comment by SkySkimmer** [comment]
Compcert failed due to https://github.com/AbsInt/CompCert/pull/275#issuecomment-467336477 
Should be fixed now

**Review by SkySkimmer — APPROVED** [review]
Not sure why bedrock failed
Linter error is spurious (#9653) and in the per-commit part, we can ignore it.

Do we merge without the .glob dependencies?

**Comment by ejgallego** [comment]
> Do we merge without the .glob dependencies?

Nope, let me fix that indeed, and rebase on top of #9648

**Comment by SkySkimmer** [comment]
I'll let @Zimmi48 as dune secondary assign.

**Comment by ejgallego** [comment]
Rebased, but I am not sure how to test the deploy job (@vbgl ?)

**Comment by vbgl** [comment]
To test deploy jobs, I’ve manually started a CI pipeline with the appropriate credentials. I can do it for you if you think its necessary.

**Comment by ejgallego** [comment]
> To test deploy jobs, I’ve manually started a CI pipeline with the appropriate credentials. I can do it for you if you think its necessary.

Yup but that would overwrite the main Coq repos, right?

I guess I'll deploy to my own repos, and make the clone target a variable then.

**Comment by vbgl** [comment]
It will not overwrite, but put you artifact in a separate directory (named as your branch: `dune+coqdoc`).

**Comment by ejgallego** [comment]
> It will not overwrite, but put you artifact in a separate directory (named as your branch: `dune+coqdoc`).

Ah Ok, cool then!

**Comment by ejgallego** [comment]
I'd appreciate indeed if you can push a build to see whether that works.

**Comment by vbgl** [comment]
Pipeline started. Result should appear in a few hours there: https://coq.github.io/doc/dune+coqdoc

**Comment by ejgallego** [comment]
> Pipeline started. Result should appear in a few hours there: https://coq.github.io/doc/dune+coqdoc

Muchas gracias @vbgl

File: Makefile.dune [hunk 5]
```diff
@@ -4,7 +4,7 @@
 .PHONY: help voboot states world watch check          # Main developer targets
 .PHONY: coq coqide coqide-server                      # Package targets
 .PHONY: quickbyte quickopt                            # Partial / quick developer targets
-.PHONY: test-suite refman-html apidoc release         # Accesory targets
+.PHONY: test-suite refman-html stdlib-html apidoc release # Accesory targets
```
Comment (Zimmi48) [inline Makefile.dune]: Probably to split this line in two, including one for Documentation targets.

File: doc/stdlib/dune [hunk 6]
```diff
@@ -0,0 +1,20 @@
+; This is an ad-hoc rule to easy the migration, it should be handled
```
Comment (Zimmi48) [inline doc/stdlib/dune]:
```suggestion
; This is an ad-hoc rule to ease the migration, it should be handled
```

**Review by Zimmi48 — APPROVED** [review]

File: Makefile.dune (hunk 5 above)
Comment (ejgallego) [reply Makefile.dune]: Indeed.

File: doc/stdlib/dune (hunk 6 above)
Comment (ejgallego) [reply doc/stdlib/dune]: Fixed thanks.

**Comment by ejgallego** [comment]
Will wait for the deploy build to finish before the rebase.

**Comment by Zimmi48** [comment]
There is also a problem with the artifacts (which should in principle make the deploy job fail): `_build/default/doc/stdlib/html/` is not exported. See https://gitlab.com/coq/coq/-/jobs/167795044/artifacts/browse/_build/default/doc/.

**Comment by ejgallego** [comment]
Well-spotted @Zimmi48 ahhh; indeed @SkySkimmer was right, I've added an extra CI job which should help consistency.

Fixed version uploaded, deploy job needs relaunch.

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -346,6 +346,15 @@ doc:refman:dune:
...
       - _build/default/doc/sphinx_build/html
 
+doc:stdlib:dune:
+  <<: *dune-ci-template
+  variables:
+    <<: *dune-ci-template-vars
+    DUNE_TARGET: stdlib-html
+  artifacts:
+    <<: *dune-ci-template-artifacts
+      - _build/default/doc/stdlib/html
```
Comment (SkySkimmer) [inline .gitlab-ci.yml]: Missing a `paths` line

Comment (ejgallego) [reply .gitlab-ci.yml]: Ups, thanks !

File: .gitlab-ci.yml (hunk 3 above)
Comment (ejgallego) [reply .gitlab-ci.yml]: I think you were right and I've changed this.

File: doc/stdlib/dune (hunk 2 above)
Comment (ejgallego) [reply doc/stdlib/dune]: Added a comment to update in the future, thanks!

**Comment by vbgl** [comment]
Pipeline launched for 7ba806bb.

**Comment by vbgl** [comment]
The deploy jobs went through (also for the last update). There are some issues with the documentation of the stdlib. E.g., at https://coq.github.io/doc/dune+coqdoc/stdlib/, in the “navigation” part, there is a dead link to `genindex.html`.

**Comment by ejgallego** [comment]
> The deploy jobs went through (also for the last update). There are some issues with the documentation of the stdlib. E.g., at https://coq.github.io/doc/dune+coqdoc/stdlib/, in the “navigation” part, there is a dead link to `genindex.html`.

It seems to work well here @vbgl , are you sure you were not looking to an older build [which had this problem]

So far, the docs looks pretty good to me.

**Comment by vbgl** [comment]
My bad. It looks good indeed.

**Review by Zimmi48 — APPROVED** [review]

**Comment by ejgallego** [comment]
Thanks to all for the help!

**Pull request merged**